TESTS/mbed_drivers/timerevent/main.cpp

```diff
@@ -0,0 +1,352 @@
+    }
+
+    TestTimerEvent(const ticker_data_t *data) :
+            TimerEvent(data), sem(1) {
+    }
+
+    virtual ~TestTimerEvent() {
+    }
+
+    void insert(timestamp_t ts) {
+        TimerEvent::insert(ts);
+    }
+
+    void insert_absolute(us_timestamp_t ts) {
+        TimerEvent::insert_absolute(ts);
+    }
+
+    void remove() {
+        TimerEvent::remove();
+    }
+
+    int32_t sem_wait(uint32_t millisec) {
+        return sem.wait(millisec);
+    }
+
+    us_timestamp_t ticker_read_us() {
+        return ::ticker_read_us(_ticker_data);
+    }
+
+    const ticker_event_t *get_event_addr() {
+        return &event;
+    }
+};
+
+class TestTimerEventRelative: public TestTimerEvent {
+public:
+    TestTimerEventRelative() :
+            TestTimerEvent() {
+    }
+
+    TestTimerEventRelative(const ticker_data_t *data) :
+            TestTimerEvent(data) {
+    }
+
+    // Set relative timestamp of internal event to present_time + ts
+    void set_future_timestamp(timestamp_t ts) {
+        insert(::ticker_read(_ticker_data) + ts);
+    }
+
+    void set_timestamp(timestamp_t ts) {
+        insert(ts);
+    }
+};
+
+class TestTimerEventAbsolute: public TestTimerEvent {
+public:
+    TestTimerEventAbsolute() :
+            TestTimerEvent() {
+    }
+
+    TestTimerEventAbsolute(const ticker_data_t *data) :
+            TestTimerEvent(data) {
+    }
+
+    // Set absolute timestamp of internal event to present_time + ts
+    void set_future_timestamp(us_timestamp_t ts) {
+        insert_absolute(::ticker_read_us(_ticker_data) + ts);
+    }
+
+    void set_timestamp(us_timestamp_t ts) {
+        insert_absolute(ts);
+    }
+};
+
+// mock ticker interface
+void mock_ticker_fun(void) {}
+void mock_ticker_set_interrupt(timestamp_t ts) {}
+uint32_t mock_ticker_read_inc(void) {
+    static uint32_t cnt;
+    return ++cnt;
+}
+uint32_t mock_ticker_read_const(void) {
+    return MOCK_TICKER_CNT;
+}
+
+const ticker_interface_t mock_ticker_interface = {
+    .init = mock_ticker_fun,
+    .read = mock_ticker_read_inc,
+    .disable_interrupt = mock_ticker_fun,
+    .clear_interrupt = mock_ticker_fun,
+    .set_interrupt = mock_ticker_set_interrupt,
+    .fire_interrupt = mock_ticker_fun
+};
+
+/** Template for tests: ticker_queue handling
+ *
+ * Test ticker_queue after @a insert call
+ * Given an instance of @a TimerEvent subclass initialized with
+ *     an empty @a ticker_queue structure
+ * When a tiestamp is set with @a insert() call
+ * Then @a ticker_queue is set properly
+ * When the timestamp is removed
+ * Then @a ticker_queue is set properly
+ *
+ * Test ticker_queue after @a insert_absolute call
+ * Given an instance of @a TimerEvent subclass initialized with
+ *     an empty @a ticker_queue structure
+ * When a tiestamp is set with @a insert_absolute() call
+ * Then @a ticker_queue is set properly
+ * When the timestamp is removed
+ * Then @a ticker_queue is set properly
+ */
+template<typename T>
+void test_ticker_queue(void) {
```

Comment:
I would say this test goes to deep into the internals and tests other components. Problem with this approach is that internals can change and API should be stable. TimerEvent only calls Tickers API so I would only test it to this depth.

Reply:
Fixed -- after the update tests don't check any inernal structures.